### PR TITLE
[SystemZ][z/OS] This change adds support for the PPA2 section in zOS

### PR DESCRIPTION
**Pull request by ysyeda**
This PR adds support for the PPA2 fields.

### CODE REVIEW

**Comment by ysyeda** [comment]
ping

File: clang/lib/Basic/LangStandards.cpp [hunk 1]
```diff
@@ -10,9 +10,49 @@
...
 #include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/FormatVariadic.h"
 #include "llvm/TargetParser/Triple.h"
 using namespace clang;
 
+const char *clang::languageToString(Language L) {
+  // I would like to make this function and the definition of Language
+  // in the .h file simply expand the contents of a .def file.
+  // However, in the .h the members of the enum have doxygen annotations
+  // and/or comments which would be lost.
```
Comment (uweigand) [inline clang/lib/Basic/LangStandards.cpp:21]: I'm not sure I have a good solution for this, but I'd like to see comments on this from the wider clang community.  Maybe it would make sense to submit just this function as a separate PR.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -2267,6 +2268,14 @@ void Clang::AddSystemZTargetArgs(const ArgList &Args,
...
     CmdArgs.push_back("soft");
   }
+
+  if (Triple.isOSzOS()) {
+    CmdArgs.push_back("-mllvm");
+    CmdArgs.push_back(
+        Args.MakeArgString(llvm::Twine("-translation-time=")
+                               .concat(llvm::Twine(std::time(nullptr)))
+                               .str()));
+  }
```
Comment (uweigand) [inline clang/lib/Driver/ToolChains/Clang.cpp:2278]: Is there a good reason for this to be a command-line option, or could this also be a module flag just like the others added above?  Seems odd to use two different mechanisms here.

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp [hunk 3]
```diff
@@ -1026,6 +1036,68 @@ void SystemZAsmPrinter::emitADASection() {
...
+  OutStreamer->pushSection();
+  OutStreamer->switchSection(getObjFileLowering().getIDRLSection());
+  constexpr unsigned IDRLDataLength = 30;
+  std::time_t Time = TranslationTime;
+
+  uint32_t ProductVersion = getProductVersion(M);
+  uint32_t ProductRelease = getProductRelease(M);
+
+  std::string ProductID;
+  if (auto *MD = M.getModuleFlag("Product Id"))
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1070]: This is never set in clang by your patch.  Should it be?

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp [hunk 4]
```diff
@@ -1296,6 +1372,124 @@ void SystemZAsmPrinter::emitPPA1(MCSymbol *FnEndSym) {
...
+  OutStreamer->emitInt8(0x04); // Control level 4 (XPLink)
+  OutStreamer->emitAbsoluteSymbolDiff(CELQSTRT, PPA2Sym, 4);
+  OutStreamer->emitInt32(0x00000000);
+  OutStreamer->emitAbsoluteSymbolDiff(DateVersionSym, PPA2Sym, 4);
+  OutStreamer->emitInt32(
+      0x00000000); // Offset to main entry point, always 0 (so says TR).
+  uint8_t Flgs = static_cast<uint8_t>(PPA2Flags::CompileForBinaryFloatingPoint);
+  Flgs |= static_cast<uint8_t>(PPA2Flags::CompiledWithXPLink);
+
+  if (auto *MD = M.getModuleFlag("zos_le_char_mode")) {
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1460]: This is also nowhere set in clang right now.  Should it be?   If it isn't set, is EBCDIC the right default?

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 5]
```diff
@@ -976,6 +976,24 @@ void CodeGenModule::Release() {
       Context.getTypeSizeInChars(Context.getWideCharType()).getQuantity();
   getModule().addModuleFlag(llvm::Module::Error, "wchar_size", WCharWidth);
 
+  if (getTriple().isOSzOS()) {
+    int32_t ProductVersion, ProductRelease, ProductPatch;
+    ProductVersion = LLVM_VERSION_MAJOR, ProductRelease = LLVM_VERSION_MINOR,
+    ProductPatch = LLVM_VERSION_PATCH;
```
Comment (uweigand) [inline clang/lib/CodeGen/CodeGenModule.cpp:982]: Do we need extra variables for this?

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 6]
```diff
@@ -976,6 +976,24 @@ void CodeGenModule::Release() {
...
+    getModule().addModuleFlag(llvm::Module::Warning, "Product Major Version",
+                              ProductVersion);
+    getModule().addModuleFlag(llvm::Module::Warning, "Product Minor Version",
+                              ProductRelease);
+    getModule().addModuleFlag(llvm::Module::Warning, "Product Patchlevel",
+                              ProductPatch);
```
Comment (uweigand) [inline clang/lib/CodeGen/CodeGenModule.cpp:988]: These are set only on z/OS, but the flag names appear generic.  If they're (set and) used only on z/OS, shouldn't they carry a z/OS specific name like below?

File: clang/lib/Basic/LangStandards.cpp (hunk 1 above)
Comment (ysyeda) [reply clang/lib/Basic/LangStandards.cpp:21]: Sure, I can create another PR for this.

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp (hunk 3 above)
Comment (ysyeda) [reply llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1070]: This is now set in clang.

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp (hunk 4 above)
Comment (ysyeda) [reply llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1460]: This is now set in clang.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 5 above)
Comment (ysyeda) [reply clang/lib/CodeGen/CodeGenModule.cpp]: No, I've removed the extra variables.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 6 above)
Comment (ysyeda) [reply clang/lib/CodeGen/CodeGenModule.cpp]: This has been changed to use the `CLANG_VERSION_MAJOR`, `CLANG_VERSION_MINOR`, etc.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (ysyeda) [reply clang/lib/Driver/ToolChains/Clang.cpp]: I refactored the translation time handling to use a module flag instead of a command line option.

**Comment by ysyeda** [comment]
@uweigand can you have another look over this PR? The[ PR ](https://github.com/llvm/llvm-project/pull/69487)for languageToString has been approved and merged, and all your remaining comments have been addressed.

File: clang/include/clang/Basic/LangOptions.def [hunk 7]
```diff
@@ -90,6 +90,7 @@ LANGOPT(C23               , 1, 0, "C23")
 LANGOPT(MSVCCompat        , 1, 0, "Microsoft Visual C++ full compatibility mode")
 LANGOPT(Kernel            , 1, 0, "Kernel mode")
 LANGOPT(MicrosoftExt      , 1, 0, "Microsoft C++ extensions")
+LANGOPT(ASCIICharMode     , 1, 1, "z/OS Language Environment Character mode")
```
Comment (uweigand) [inline clang/include/clang/Basic/LangOptions.def:93]: This doesn't appear to do anything except for setting the PPA flag.   Is this correct?  If the implementation is simply incomplete / not present yet, I don't think we should add this flag now - rather, we should have the PPA flag reflect the default and update this once full support for ASCII mode is merged.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 8]
```diff
@@ -976,6 +976,40 @@ void CodeGenModule::Release() {
       Context.getTypeSizeInChars(Context.getWideCharType()).getQuantity();
   getModule().addModuleFlag(llvm::Module::Error, "wchar_size", WCharWidth);
 
+  if (getTriple().isOSzOS()) {
+    getModule().addModuleFlag(llvm::Module::Warning, "Product Major Version",
+                              uint32_t(CLANG_VERSION_MAJOR));
+    getModule().addModuleFlag(llvm::Module::Warning, "Product Minor Version",
+                              uint32_t(CLANG_VERSION_MINOR));
+    getModule().addModuleFlag(llvm::Module::Warning, "Product Patchlevel",
+                              uint32_t(CLANG_VERSION_PATCHLEVEL));
```
Comment (uweigand) [inline clang/lib/CodeGen/CodeGenModule.cpp:985]: I think I commented on that previously:  these flags are set only on z/OS, but have rather generic names.   If they're z/OS specific, I think they should have names in the "zos_..." name space like the others.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 9]
```diff
@@ -976,6 +976,40 @@ void CodeGenModule::Release() {
...
+
+    std::string ProductId;
+#ifdef CLANG_VENDOR
+    ProductId = #CLANG_VENDOR;
+#else
+    ProductId = "clang";
+#endif
+
+    getModule().addModuleFlag(llvm::Module::Error, "Product Id",
+                              llvm::MDString::get(VMContext, ProductId));
```
Comment (uweigand) [inline clang/lib/CodeGen/CodeGenModule.cpp:1001]: Same comment as above about the name space.  Also, please move this up next to the other product-related flags.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 10]
```diff
@@ -976,6 +976,40 @@ void CodeGenModule::Release() {
...
+
+    getModule().addModuleFlag(llvm::Module::Error, "TranslationTime",
```
Comment (uweigand) [inline clang/lib/CodeGen/CodeGenModule.cpp:1003]: I don't think `Error` is the correct setting for this.  That would prevent LTO unless modules were built at the exact same time.  Rather, I think this should be `Max` here.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 11]
```diff
@@ -1765,7 +1765,7 @@ void Clang::RenderTargetOptions(const llvm::Triple &EffectiveTriple,
     break;
 
   case llvm::Triple::systemz:
-    AddSystemZTargetArgs(Args, CmdArgs);
+    AddSystemZTargetArgs(EffectiveTriple, Args, CmdArgs);
```
Comment (uweigand) [inline clang/lib/Driver/ToolChains/Clang.cpp:1768]: This (and the related) changes are no longer needed now.

**Review by uweigand — CHANGES_REQUESTED** [review]

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 8 above)
Comment (ysyeda) [reply clang/lib/CodeGen/CodeGenModule.cpp:985]: Done. All your remaining comments have also been addressed.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 12]
```diff
@@ -976,6 +976,37 @@ void CodeGenModule::Release() {
...
+                              llvm::MDString::get(VMContext, ProductId));
+
+    // Record the language because we need it for the PPA2.
+    StringRef lang_str = languageToString(
+        LangStandard::getLangStandardForKind(LangOpts.LangStd).Language);
+    getModule().addModuleFlag(llvm::Module::Error, "zos_cu_language",
+                              llvm::MDString::get(VMContext, lang_str));
+
+    getModule().addModuleFlag(llvm::Module::Max, "zos_translation_time",
+                              static_cast<uint64_t>(std::time(nullptr)));
```
[the remark this answers is not in the document]
Comment (ysyeda) [reply clang/lib/CodeGen/CodeGenModule.cpp]: I added a change to prefer `SourceDateEpoch` if it is set, and `std::time(nullptr)` otherwise.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 13]
```diff
@@ -976,6 +976,37 @@ void CodeGenModule::Release() {
...
+    ProductId = "clang";
+#endif
+    getModule().addModuleFlag(llvm::Module::Error, "zos_product_id",
+                              llvm::MDString::get(VMContext, ProductId));
+
+    // Record the language because we need it for the PPA2.
+    StringRef lang_str = languageToString(
+        LangStandard::getLangStandardForKind(LangOpts.LangStd).Language);
+    getModule().addModuleFlag(llvm::Module::Error, "zos_cu_language",
+                              llvm::MDString::get(VMContext, lang_str));
```
[the remark this answers is not in the document]
Comment (ysyeda) [reply clang/lib/CodeGen/CodeGenModule.cpp:1001]: We want to keep this as llvm::Module::Error. Merging of PPA2's is not defined (a limitation of the system). We will have some limitations with LTO in the future, e.g. mixing of C and C++, etc. will exclude the use of LTO. For now that should be fine.

**Comment by ysyeda** [comment]
ping

**Comment by ysyeda** [comment]
@uweigand your comments have been addressed. Can you please give this another review?

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp [hunk 14]
```diff
@@ -1026,6 +1030,78 @@ void SystemZAsmPrinter::emitADASection() {
   OutStreamer->popSection();
 }
 
+static uint32_t getProductVersion(Module &M) {
+  if (auto *VersionVal = cast_or_null<ConstantAsMetadata>(
+          M.getModuleFlag("zos_product_major_version")))
+    return cast<ConstantInt>(VersionVal->getValue())->getZExtValue();
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1036]: I think instead of these two casts, these days you're supposed to use the new `mdconst` interface, e.g. like
```
if (auto *VersionVal = mdconst::extract_or_null<ConstantInt>(
        M.getModuleFlag("zos_product_major_version"))
```

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp [hunk 15]
```diff
@@ -1026,6 +1030,78 @@ void SystemZAsmPrinter::emitADASection() {
...
+  uint32_t ProductRelease = getProductRelease(M);
+
+  std::string ProductID;
+  if (auto *MD = M.getModuleFlag("zos_product_id"))
+    ProductID = cast<MDString>(MD)->getString().str();
+
+  if (ProductID.empty()) {
+    char ProductIDFormatted[11]; // 10 + null.
+    snprintf(ProductIDFormatted, sizeof(ProductIDFormatted), "LLVM  %02d%02d",
+             ProductVersion, ProductRelease);
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1080]: In this case, `ProductVersion` and `ProductRelease` seem to be present twice, first as part of the ID and then again separately.  (This is different from when clang sets the ID as just "clang".)  Is this deliberate?

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp [hunk 16]
```diff
@@ -1026,6 +1030,78 @@ void SystemZAsmPrinter::emitADASection() {
...
+    ProductID = ProductIDFormatted;
+  }
+
+  // Remove - from Product Id, which makes it consistent with legacy.
+  // The binder expects alphanumeric characters only.
+  std::size_t DashFound = ProductID.find("-");
+  if (DashFound != std::string::npos)
+    ProductID.erase(ProductID.begin() + DashFound);
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1088]: Can there be multiple dashes, and if so, should they all be removed?   What about other disallowed characters?  At this point the ProductID could be user-provided in the IR, so shouldn't be careful here?

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp [hunk 17]
```diff
@@ -1026,6 +1030,78 @@ void SystemZAsmPrinter::emitADASection() {
...
+
+  SmallString<IDRLDataLength + 1> TempStr;
+  raw_svector_ostream O(TempStr);
+  O << formatv("{0}{1,0-2:d}{2,0-2:d}{3:%Y-%m-%d %H:%M:%S}", ProductID.c_str(),
+               ProductVersion, ProductRelease, llvm::sys::toUtcTime(Time));
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1093]: Can this overflow?  In the default case above, you ensure the ProductID has 10 bytes.  But in the case where the ID comes from the IR, its length can be anything, and is not checked here.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 18]
```diff
@@ -976,6 +976,40 @@ void CodeGenModule::Release() {
...
+
+    time_t TT = PreprocessorOpts.SourceDateEpoch
+                    ? *PreprocessorOpts.SourceDateEpoch
+                    : std::time(nullptr);
+    getModule().addModuleFlag(llvm::Module::Max, "zos_translation_time",
+                              static_cast<uint64_t>(TT));
+
+    getModule().addModuleFlag(llvm::Module::Error, "zos_le_char_mode",
+                              llvm::MDString::get(VMContext, "ascii"));
```
Comment (uweigand) [inline clang/lib/CodeGen/CodeGenModule.cpp:1010]: Is "ascii" the correct default here?   In either case, please add a comment stating multiple modes will be supported in the future, so we understand why this code is here.

**Review by uweigand — COMMENTED** [review]
Please find a couple of additional comments inline, related to module flag processing.

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp (hunk 15 above)
Comment (ysyeda) [reply llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp]: the ID has been fixed to be just "LLVM".

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp (hunk 17 above)
Comment (ysyeda) [reply llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp]: I added a check for this.

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp (hunk 16 above)
Comment (ysyeda) [reply llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp]: There wouldn't be multiple dashes, and the comment that the binder expects alphanumeric characters only was incorrect and has been removed. This piece of code has been moved to CodeGenModule.cpp.

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp [hunk 19]
```diff
@@ -1026,6 +1030,71 @@ void SystemZAsmPrinter::emitADASection() {
...
+  uint32_t ProductVersion = getProductVersion(M);
+  uint32_t ProductRelease = getProductRelease(M);
+
+  std::string ProductID;
+  if (auto *MD = M.getModuleFlag("zos_product_id"))
+    ProductID = cast<MDString>(MD)->getString().str();
+
+  if (ProductID.empty()) {
+    char ProductIDFormatted[11]; // 10 + null.
+    snprintf(ProductIDFormatted, sizeof(ProductIDFormatted), "LLVM");
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1079]: This `snprintf` call now looks superfluous - just set ProductID to "LLVM".

Also, I think this should all now go into a `getProductID` helper analogous to `getProductVersion` and the like.

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp (hunk 16 above)
Comment (uweigand) [reply llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp]: In the back-end, the string comes from a module flag in the IR - when providing their own IR file, a user could set that flag to anything, including a string with multiple dashes.  So if this is a problem for the rest of the toolchain, we need to be more careful in handling it.

If this is not actually a problem in the binder, then I guess removing that code here is fine.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 20]
```diff
@@ -976,6 +976,46 @@ void CodeGenModule::Release() {
...
+    std::string ProductId;
+#ifdef CLANG_VENDOR
+    ProductId = #CLANG_VENDOR;
+#else
+    ProductId = "clang";
+#endif
+    // Remove - from Product Id, which makes it consistent with legacy.
+    std::size_t DashFound = ProductId.find("-");
+    if (DashFound != std::string::npos)
+      ProductId.erase(ProductId.begin() + DashFound);
```
Comment (uweigand) [inline clang/lib/CodeGen/CodeGenModule.cpp:997]: Now that this code is moved here, it could only make a difference when compiling clang with a `CLANG_VENDOR` containing a dash.  If that is a problem for your use case, then why not simply build with the desired `CLANG_VENDOR` to begin with?  It's a bit unclear to me which general problem this code solves at this place now.

Comment (perry-ca) [reply clang/lib/CodeGen/CodeGenModule.cpp:997]: The proper product id contains the dash.  The only place we don't include the dash is in the PPA2 timestamp string.  This is to be compat with XL.   We would like all uses of CLANG_VENDOR to use the proper product id (except for this one use in PPA2).

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp (hunk 16 above)
Comment (ysyeda) [reply llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp]: I tried changing the productID to an arbitrary string containing multiple dashes and non alphanumerics, and it does not cause issues in the binder. The original comment is wrong and the only need to remove the '-' is for the C/C++ compiler to match xlc/xlC. So this code can be removed from here.

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp (hunk 19 above)
Comment (ysyeda) [reply llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp]: Done

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp [hunk 21]
```diff
@@ -1026,6 +1030,71 @@ void SystemZAsmPrinter::emitADASection() {
   OutStreamer->popSection();
 }
 
+static uint32_t getProductVersion(Module &M) {
+  if (auto *VersionVal = mdconst::extract_or_null<ConstantInt>(
+          M.getModuleFlag("zos_product_major_version")))
+    return VersionVal->getValue().getZExtValue();
```
[the remark this answers is not in the document]
Comment (ysyeda) [reply llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp]: Done

File: llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp [hunk 22]
```diff
@@ -1026,6 +1030,72 @@ void SystemZAsmPrinter::emitADASection() {
...
+  O << formatv("{0}{1,0-2:d}{2,0-2:d}{3:%Y-%m-%d %H:%M:%S}",
+               ProductID.substr(0, 10).c_str(), ProductVersion, ProductRelease,
+               llvm::sys::toUtcTime(Time));
+  SmallString<IDRLDataLength> Data;
+  ConverterEBCDIC::convertToEBCDIC(TempStr, Data);
+
+  OutStreamer->emitInt8(0);               // Reserved.
+  OutStreamer->emitInt8(3);               // Format.
+  OutStreamer->emitInt16(IDRLDataLength); // Length.
+  OutStreamer->emitBytes(Data.str());
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:1095]: One more question about this string - the Length field is hardcoded to 30.  Does that mean that there are exactly 30 bytes supposed to always follow here?   I'm not sure this is guaranteed by the formatv above ...

Also, there doesn't seem to be any test that verifies the layout of this IDRL section, I think we definitely need one.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 20 above)
Comment (ysyeda) [reply clang/lib/CodeGen/CodeGenModule.cpp]: I removed this from this PR as it's not required for the binder.

**Review by uweigand — APPROVED** [review]
This version LGTM now, thanks!

**Comment by dyung** [comment]
@ysyeda the test you added, `clang/test/CodeGen/SystemX/systemz-ppa2.c` fails if the SystemZ target (I'm guessing) is not available. Can you add the proper `REQUIRES` line or fix the test to work without that architecture present?

https://lab.llvm.org/buildbot/#/builders/139/builds/54293
```
RUN: at line 13: /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/clang -cc1 -internal-isystem /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/lib/clang/18/include -nostdsysteminc -triple s390x-ibm-zos -xc -S -o - /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/llvm-project/clang/test/CodeGen/SystemZ/systemz-ppa2.c | /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/FileCheck /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/llvm-project/clang/test/CodeGen/SystemZ/systemz-ppa2.c --check-prefix CHECK-C
+ /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/clang -cc1 -internal-isystem /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/lib/clang/18/include -nostdsysteminc -triple s390x-ibm-zos -xc -S -o - /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/llvm-project/clang/test/CodeGen/SystemZ/systemz-ppa2.c
+ /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/FileCheck /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/llvm-project/clang/test/CodeGen/SystemZ/systemz-ppa2.c --check-prefix CHECK-C
error: unable to create target: 'No available targets are compatible with triple "s390x-ibm-zos"'
```

File: llvm/test/CodeGen/SystemZ/zos-ppa2.ll [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+; CHECK:    .ascii  "\361\371\367\360\360\361\360\361\360\360\360\360\360\360"
+; CHECK:    .short  0
+; CHECK:    .quad   @@PPA2-CELQSTRT                 * A(PPA2-CELQSTRT)
+; CHECK: @@PPA1_void_test_0:
+; CHECK:    .long   @@PPA2-@@PPA1_void_test_0       * Offset to PPA2
+; CHECK:    .section    "B_IDRL"
+; CHECK:    .byte   0
+; CHECK:    .byte   3
+; CHECK:    .short  30
+; CHECK:    .ascii  "\323\323\345\324@@@@@@\361\370\360\360\361\371\367\360\360\361\360\361\360\360\360\360\360\360\360\360"
```
Comment (tstellar) [inline llvm/test/CodeGen/SystemZ/zos-ppa2.ll:27]: Hi,

Based on 4236d56141be3a421d9e7fe8bb7f1ba1663ec97a, it appears that this test encodes the llvm version into the test.  Is there some way to change this test so it does not container the llvm version in it.  It's not practical to have to update this every time there is a release.

Comment (dyung) [reply llvm/test/CodeGen/SystemZ/zos-ppa2.ll:27]: I just posted a PR to make the version check a regex if the author accepts it. https://github.com/llvm/llvm-project/pull/79528
